design/httpproxy-conflict.md

```diff
@@ -0,0 +1,82 @@
+# HTTPProxy Conflict Resolution Design
+
+Status: Draft
+
+## Abstract
+
+Configuration errors in HTTPProxy resources can result in objects being marked as invalid and conversely will not serve traffic.
+These errors can come from invalid route configuration, missing service references, or invalid includes to other resources. 
+
+## Background
+HTTPProxy is designed as a multi-team aware Ingress, reduce annotation sprawl by providing a sensible home for configuration items.
+Additionally, HTTPProxy offers a feedback loop to users via a `Status` field on the object.
+Contour has further extended this to add `Conditions` which allows for a set of `Errors` and `Warnings` which inform users of errors in the HTTPProxy resource.
+
+When the first error or warning is detected, these fields are updated and in the status of the HTTPProxy object, however if there are more errors, they are not all shown to the user, only the first.
+There are ongoing efforts to change this logic to inform of all the errors or warnings dealing with a specific object.
+
+Contour currently invalidates the entire resource if a configuration error is encountered which can cause downtime for specific routes. 
+
+## Goals
+- Serve valid traffic configurations if portions of the spec are invalid
+
+## Non Goals
+- Surface a complete set errors or warnings for a resource (this is a different issue)
+
+## High-Level Design
+Contour will set the error or warning condition when an error is encountered, but do its best to still serve valid configurations.
+
+- If an invalid route is encountered, Contour will return a `404` HTTP status code to the requester
+- If an invalid service is encountered,Contour will return a`503` HTTP status code to the requester
+
+It's difficult to find errors and block them from breaking valid configurations due to how Contour currently processes objects.
+Contour takes a passive look to the cluster by reacting to changes after they have changed. 
+Due to this, Contour cannot block changes to resources before they are committed like an `Admission Controller` would be able to do (See alternatives: `HTTPProxyInstance`)
+
+## Detailed Design
+
+### Invalid Include
+Contour utilizes an inclusion model to allow for proper route/header delegation across namespaces and teams within a Kubernetes cluster.
+If an include does not have a matching HTTPProxy.Name or HTTPProxy.Namespace, then Contour will set an error in the Conditions of the object.
+Any requests matching the `Conditions` on the `Spec.Include` will return a `404` to the requester.
+
+If the include was valid previously and is now no longer valid, the routes will no longer be valid which the include previously enabled.
+
+The logic described above should only apply if the object has a proper delegation chain to the defined `Conditions` on the include. 
+If an object creates an `Include` to values that it does not have permission to, then it shouldn't create `404` responses.
```

Comment:
I agree that "access denied" should be treated the same as "object not found" in general terms, preferably with a status message somewhere that tells you it can't be found because access denied.

It would be possible to use this feature for object enumeration, by trying various names and seeing which ones are "not found" and which are "access denied", but that seems like an okay price to pay - I don't think that knowing the names of objects in a namespace you don't have access to is an unacceptable risk.

Reply:
I removed this, I was pulling in IngressRoute logic for some reason. Anyone who can create an Include will only be able to do so at any level they have access to, do there's no way for someone to request a path they don't have access to.